src/main/java/net/coderbot/iris/shaderpack/DefineOptionParser.java

```diff
@@ -0,0 +1,332 @@
+package net.coderbot.iris.shaderpack;
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.StringUtils;
+
+import net.minecraft.util.Util;
+
+public class DefineOptionParser {
+	/*
+	 Regex for matching boolean options
+	  Match if or if not the line starts with anynumber of backslashes that are more than 2 ("//")
+	  Match 0 or more whitspace after the "//"
+	  Match the #define
+	  Match 1 whitespace after that
+	  Match any letter, number, or underscore name
+	  Match 0 or more whitespace after that
+	  Match any comments on the option after that
+	 */
+	private static final Pattern BOOLEAN_OPTION_PATTERN = Pattern.compile("(?<startingComment>//+)?\\s*(?<define>#define)\\s+(?<name>\\w+)\\s*(?<comment>(?<commentChar>//+)(?<commentContent>.*))?");
+	/*
+	  Regex for matching ifdef patterns for boolean options
+	  Match the ifdef or ifndef keyword
+	  match whitespace that must be there
+	  match a word that is the name of the keyword
+	 */
+	private static final Pattern IFDEF_IFNDEF_PATTERN = Pattern.compile("(?<ifdef>#ifdef|#ifndef)\\s+(?<name>\\w+)");
+	/*
+	   Regex that matches for integer and float options
+	   Match 1 or more whitespace after #define
+	   match a word (name of the option). Put in parenthesis for grouping
+	   match 1 or more whitespace after the name
+	      match a negative if there
+	      match f or F float keyword
+	      match 0-9
+	      match a "." for floats
+	   match 1 or more of the char group
+	   match 0 or more whitespace
+	   match if there is a comment following the line or not
+	 */
+	private static final Pattern FLOAT_INTEGER_OPTION_PATTERN = Pattern.compile("(?<define>#define)\\s+(?<name>\\w+)\\s+(?<value>-?[\\d.fF]+)\\s*(?<comment>(?<commentChar>//+)(?<commentContent>.*))?");
+	/*
+	 Regex that matches for only integers and not floats
+	 Same as above but in the char class we remove the float specific checks
+		remove matching a "."
+		remove matching a "f" or a "F"
+	 */
+	private static final Pattern INTEGER_OPTION_PATTERN = Pattern.compile("(?<define>#define)\\s+(?<name>\\w+)\\s+(?<value>-?\\d+)\\s*(?<comment>(?<commentChar>//+)(?<commentContent>.*))?");
+	/*
+	Some shaderpacks like sildurs have #define directives that are named with the program name
+	like #define gbuffers_textured
+	optifine does not use these in their config so we will not as well
+	TODO figure what optifine does with these program defined names and implement
+	 */
+	private static final Set<String> IGNORED_PROGRAM_NAMES = Util.make(new HashSet<>(), (set) -> {
```

Comment:
It seems that optifine does parse these as options, but does not, as you said, add it to the config screen without specification. With a modified shaders.properties, the option is shown on Optifine (See picture below). I think the best thing to do would be to do what Optifine does and parse these options, but as FoundationGames said, not include all these options in the config file, and instead only include options that have changed their value. 
![image](https://user-images.githubusercontent.com/69648849/117583045-dd1c7480-b0b9-11eb-94f0-627ba5b107dc.png)

Reply:
However, if we did make only the changed options be put in the config file, this would require the config gui to be in place or otherwise the config would not be editable at all

Reply:
Makes sense, good to hear. I think we should only include changed options in the config for now, and then in the near term we can either get a config GUI set up, or have an easy way to list all settings with default values. I think it would be completely fine if this PR just implemented the "backend" code of shader configuration and left the part of making it easily editable to a different PR.

Reply:
Okay sounds good!